File: horse.py
from pyray import Vector2, Rectangle
import pyray as rl


class Horse:
    def __init__(
        self, name: str, color: rl.Color, boop: rl.Sound, texture: str
    ) -> None:
        self.name: str = name
        self.position: Vector2 = Vector2(0, 0)
        self.radius: int = 20
        self.color: rl.Color = color
        self.speed: Vector2 = Vector2(0, 0)
        self.boop: rl.Sound = boop
        self.texture: rl.Texture = rl.load_texture(texture)

    def accelerate(self) -> None:
        self.position.x += self.speed.x
        self.position.y += self.speed.y

    def collide_with_border(self, b: Rectangle):
        center_x = b.x + b.width / 2
        center_y = b.y + b.height / 2

        dx = self.position.x - center_x
        dy = self.position.y - center_y
        if abs(dx / b.width) > abs(dy / b.height):
            self.speed.x *= -1.0
            if dx > 0:
                self.position.x = b.x + b.width + self.radius
            else:
                self.position.x = b.x - self.radius
        else:
            self.speed.y *= -1.0
            if dy > 0:
                self.position.y = b.y + b.height + self.radius
            else:
                self.position.y = b.y - self.radius

    def collide_with_horse(self, h: "Horse") -> None:
        dx = self.position.x - h.position.x
        dy = self.position.y - h.position.y

        # type: ignore
        distance: float = (dx * dx + dy * dy) ** 0.5
        min_distance = self.radius + h.radius

        if distance < min_distance and distance > 0:
            # Normaliza el vector de colisión
            nx = dx / distance
            ny = dy / distance

            # Separa los círculos
            overlap = min_distance - distance
            self.position.x += nx * overlap
            self.position.y += ny * overlap

            # Refleja la velocidad del círculo 1
            # Producto punto: v · n
            dot = self.speed.x * nx + self.speed.y * ny

            # Nueva velocidad: v' = v - 2(v·n)n
            self.speed.x -= 2 * dot * nx
            self.speed.y -= 2 * dot * ny

            # Si el círculo 2 también se mueve (opcional)
            dot2 = h.speed.x * nx + h.speed.y * ny
            h.speed.x -= 2 * dot2 * nx
            h.speed.y -= 2 * dot2 * ny

    def check_collision_borders(self, bounds: list[rl.Rectangle]) -> None:
        for b in bounds:
            if rl.check_collision_circle_rec(self.position, self.radius, b):
                self.collide_with_border(b)
                rl.play_sound(self.boop)

    def check_collision_horses(self, bounds: list["Horse"]) -> None:
        for h in bounds:
            if rl.check_collision_circles(
                self.position, self.radius, h.position, h.radius
            ):
                self.collide_with_horse(h)

    def render(self) -> None:
        rl.draw_texture_ex(
            self.texture,
            Vector2(
                self.position.x - self.radius,
                self.position.y - self.radius,
            ),
            0,
            self.texture.width / 6000.0,
            rl.WHITE,
        )

        rl.draw_text(
            self.name,
            int(self.position.x) - 15,
            int(self.position.y + self.radius / 2) + 10,
            8,
            rl.BLACK,
        )

File: src/horse.cpp
#include "horse.h"

#include "raylib.h"
#include "raymath.h"
#include <print>
#include <utility>

using std::string;

Horse::Horse(string n, const string &t)
    : name{ std::move(n) }
    , texture(LoadTexture(t.c_str()))
    , radius(20) {
}

void Horse::render() {
    int text_width = MeasureText(name.c_str(), 8);

    DrawTextureEx(texture,
                  Vector2{ position.x - radius, position.y - radius },
                  0,
                  texture.width / 6000.0,
                  WHITE);
    DrawText(name.c_str(),
             position.x - text_width / 2.0f,
             position.y + radius / 2.0f + 10,
             8,
             BLACK);
}

void Horse::portrait_render(Vector2 pos) {
    DrawTextureEx(texture,
                  pos,
                  0,
                  texture.width / 3000.0,
                  WHITE);
    DrawText(name.c_str(),
             pos.x + 86,
             pos.y,
             10,
             BLACK);
}

void Horse::accelerate() {
    position.x += speed.x;
    position.y += speed.y;
}

bool Horse::collide_with_border(Rectangle& b) {
    if (!CheckCollisionCircleRec(position, radius, b)) {
        return false;
    }

    int center_x = b.x + b.width / 2;
    int center_y = b.y + b.height / 2;

    int dx = position.x - center_x;
    int dy = position.y - center_y;

    if (abs(dx / b.width) > abs(dy / b.height)) {
        speed.x *= -1.0;
        if (dx > 0) {
            position.x = b.x + b.width + radius;
        } else {
            position.x = b.x - radius;
        }
    } else {
        speed.y *= -1.0;
        if (dy > 0) {
            position.y = b.y + b.height + radius;
        } else {
            position.y = b.y - radius;
        }
    }
    return true;
}

bool Horse::collide_with_horse(Horse* h) {
    Vector2 diff = Vector2Subtract(h->position, position);
    float dist = Vector2Length(diff);
    float min_dist = radius + h->radius;

    if (dist < min_dist && dist > 0.0f) {
        Vector2 normal = Vector2Scale(diff, 1.0f / dist);

        float overlap = (min_dist - dist) / 2.0f;
        position = Vector2Subtract(position, Vector2Scale(normal, overlap));
        h->position = Vector2Add(h->position, Vector2Scale(normal, overlap));

        Vector2 relative_velocity = Vector2Subtract(h->speed, speed);
        float v_along_normal = Vector2DotProduct(relative_velocity, normal);

        if (v_along_normal > 0) {
            return false;
        }

        Vector2 impulse = Vector2Scale(normal, v_along_normal);

        speed = Vector2Add(speed, impulse);
        h->speed = Vector2Subtract(h->speed, impulse);

        return true;
    }

    return false;
}

void Horse::swap_texture(const string &texture_path) {
    UnloadTexture(texture);
    LoadTexture(texture_path.c_str());
}

Horse::~Horse() {
    std::println("{:} ha sido liberado", name);
    UnloadTexture(texture);
}
